src/source/Crypto/Dtls_openssl.c

```diff
@@ -104,24 +104,48 @@ STATUS createCertificateAndKey(INT32 certificateBits, BOOL generateRSACertificat
 {
     ENTERS();
     STATUS retStatus = STATUS_SUCCESS;
-    BIGNUM* pBne = NULL;
-    RSA* pRsa = NULL;
     X509_NAME* pX509Name = NULL;
+    UINT64 certSn;
+#if (OPENSSL_VERSION_NUMBER >= 0x30000000L)
+    EVP_PKEY_CTX* pctx = NULL;
+#else
+    RSA* pRsa = NULL;
+    BIGNUM* pBne = NULL;
     UINT32 eccGroup = 0;
     EC_KEY* eccKey = NULL;
-    UINT64 certSn;
-
+#endif
     CHK(ppCert != NULL && ppPkey != NULL, STATUS_NULL_ARG);
     CHK((*ppPkey = EVP_PKEY_new()) != NULL, STATUS_CERTIFICATE_GENERATION_FAILED);
     CHK_STATUS(dtlsFillPseudoRandomBits((PBYTE) &certSn, SIZEOF(UINT64)));
 
+#if (OPENSSL_VERSION_NUMBER >= 0x30000000L)
+    DLOGI("Detected openssl version greater than 3.0.0");
+    if (generateRSACertificate) {
+        DLOGI("Using RSA");
+        CHK_ERR(pctx = EVP_PKEY_CTX_new_from_name(NULL, "RSA", NULL), STATUS_CERTIFICATE_GENERATION_FAILED, "Error creating EVP_PKEY_CTX for RSA");
+        CHK_ERR(EVP_PKEY_keygen_init(pctx) > 0, STATUS_CERTIFICATE_GENERATION_FAILED, "Error initializing RSA keygen");
+        CHK_ERR(EVP_PKEY_CTX_set_rsa_keygen_bits(pctx, certificateBits) > 0, STATUS_CERTIFICATE_GENERATION_FAILED, "Error setting RSA key size");
+        CHK_ERR(EVP_PKEY_keygen(pctx, ppPkey) > 0, STATUS_CERTIFICATE_GENERATION_FAILED, "RSA Key generation errored");
+    } else {
+        DLOGI("Using EC");
+        CHK_ERR(pctx = EVP_PKEY_CTX_new_from_name(NULL, "EC", NULL), STATUS_CERTIFICATE_GENERATION_FAILED, "Error creating EVP_PKEY_CTX for EC");
+        CHK_ERR(EVP_PKEY_keygen_init(pctx) > 0, STATUS_CERTIFICATE_GENERATION_FAILED, "Error initializing EC keygen");
+        CHK_ERR(EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pctx, OBJ_sn2nid("prime256v1")) > 0, STATUS_CERTIFICATE_GENERATION_FAILED,
+                "Error setting curve name");
+        CHK_ERR(EVP_PKEY_keygen(pctx, ppPkey) > 0, STATUS_CERTIFICATE_GENERATION_FAILED, "EC Key generation errored");
+    }
+    EVP_PKEY_CTX_free(pctx);
+#else
     if (generateRSACertificate) {
+        DLOGI("Detected older version");
         CHK((pBne = BN_new()) != NULL, STATUS_CERTIFICATE_GENERATION_FAILED);
         CHK(BN_set_word(pBne, KVS_RSA_F4) != 0, STATUS_CERTIFICATE_GENERATION_FAILED);
-
         CHK((pRsa = RSA_new()) != NULL, STATUS_CERTIFICATE_GENERATION_FAILED);
         CHK(RSA_generate_key_ex(pRsa, certificateBits, pBne, NULL) != 0, STATUS_CERTIFICATE_GENERATION_FAILED);
         CHK((EVP_PKEY_assign_RSA(*ppPkey, pRsa)) != 0, STATUS_CERTIFICATE_GENERATION_FAILED);
+        if (pBne != NULL) {
+            BN_free(pBne);
+        }
```

Comment:
This still should be in CleanUp -- all the above CHKs if they fail will skip this free call.

Reply:
My bad. Missed moving that to cleanup when I moved the bne out of this block.